From a0f7918a196df58ba75bde88eacb676d1e17b361 Mon Sep 17 00:00:00 2001
From: Michael Ellerman <mpe@ellerman.id.au>
Date: Wed, 10 Jan 2018 22:47:40 +1100
Subject: [PATCH 5/6] powerpc/rfi-flush: Move the logic to avoid a redo into
 the sysfs code
Patch-mainline: Submitted, 2018-01-09 LKML + linuxppc-dev
References: bsc#1075088

rfi_flush_enable() includes a check to see if we're already
enabled (or disabled), and in that case does nothing.

But that means calling setup_rfi_flush() a 2nd time doesn't actually
work, which is a bit confusing.

Move that check into the sysfs code, where it really belongs.

Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>
[mauricfo: sysfs.c hunk 1: update context lines]
Signed-off-by: Mauricio Faria de Oliveira <mauricfo@linux.vnet.ibm.com>
Reviewed-by: Torsten Duwe <duwe@suse.de>
---
 arch/powerpc/kernel/setup_64.c | 3 ---
 arch/powerpc/kernel/sysfs.c    | 9 +++++++--
 2 files changed, 7 insertions(+), 5 deletions(-)

diff --git a/arch/powerpc/kernel/setup_64.c b/arch/powerpc/kernel/setup_64.c
index 88912ce55c08..5deac9992ec8 100644
--- a/arch/powerpc/kernel/setup_64.c
+++ b/arch/powerpc/kernel/setup_64.c
@@ -704,9 +704,6 @@ static void do_nothing(void *unused)
 
 void rfi_flush_enable(bool enable)
 {
-	if (rfi_flush == enable)
-		return;
-
 	if (enable) {
 		do_rfi_flush_fixups(enabled_flush_types);
 		on_each_cpu(do_nothing, NULL, 1);
diff --git a/arch/powerpc/kernel/sysfs.c b/arch/powerpc/kernel/sysfs.c
index 888d5105eb17..e8238b8c6d0a 100644
--- a/arch/powerpc/kernel/sysfs.c
+++ b/arch/powerpc/kernel/sysfs.c
@@ -211,6 +211,7 @@ static ssize_t __used store_rfi_flush(struct sysdev_class *class,
 				      struct sysdev_class_attribute *attr, const char *buf,
 				      size_t count)
 {
+	bool enable;
 	int val;
 	int ret = 0;
 
@@ -219,12 +220,16 @@ static ssize_t __used store_rfi_flush(struct sysdev_class *class,
 		return -EINVAL;
 
 	if (val == 1)
-		rfi_flush_enable(true);
+		enable = true;
 	else if (val == 0)
-		rfi_flush_enable(false);
+		enable = false;
 	else
 		return -EINVAL;
 
+	/* Only do anything if we're changing state */
+	if (enable != rfi_flush)
+		rfi_flush_enable(enable);
+
 	return count;
 }
 
-- 
2.14.3

